** Grupo:
Luana Sousa Matos - RM552621
Matheus de Freitas Silva - RM552602
Rodrigo Mendes de Almeida Silva - RM553448

In [2]:
!pip install matplotlib seaborn scikit-learn pandas


In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

Depois de instalar e importar o básico, começamos a tratar os dados. Na verdade, percebemos depois que não usamos tudo que importamos, mas deixamos do jeito que estava. 

In [4]:
df = pd.read_csv('cogumelos_MODIFICADO.csv')
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54035 entries, 0 to 54034
Data columns (total 11 columns):
 #   Column                                                  Non-Null Count  Dtype  
---  ------                                                  --------------  -----  
 0   comestivel                                              52954 non-null  float64
 1   diâmetro do chapéu                                      52954 non-null  float64
 2   forma do chapéu                                         52954 non-null  float64
 3   fixação das lamelas                                     52954 non-null  float64
 4   cor das lamelas                                         52954 non-null  float64
 5   coluna completamente inutil (me remova)                 52954 non-null  float64
 6   altura do caule                                         52954 non-null  float64
 7   largura do caule                                        52954 non-null  float64
 8   cor do caule                        

Aqui pegamos os dados, começamos a dar uma olhada no que tínhamos em mãos para entender o volume e a estrutura do csv. 

In [5]:
df.drop(['coluna completamente inutil (me remova)'], axis=1, inplace=True)
df.drop(['coluna de origem duvidosa (use por sua conta em risco)'], axis=1, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54035 entries, 0 to 54034
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   comestivel           52954 non-null  float64
 1   diâmetro do chapéu   52954 non-null  float64
 2   forma do chapéu      52954 non-null  float64
 3   fixação das lamelas  52954 non-null  float64
 4   cor das lamelas      52954 non-null  float64
 5   altura do caule      52954 non-null  float64
 6   largura do caule     52954 non-null  float64
 7   cor do caule         52954 non-null  float64
 8   estação              52954 non-null  float64
dtypes: float64(9)
memory usage: 3.7 MB


Aqui tiramos as colunas obviamente desnecessárias (obrigado professor!)

In [6]:
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 45054 entries, 0 to 54034
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   comestivel           45054 non-null  float64
 1   diâmetro do chapéu   45054 non-null  float64
 2   forma do chapéu      45054 non-null  float64
 3   fixação das lamelas  45054 non-null  float64
 4   cor das lamelas      45054 non-null  float64
 5   altura do caule      45054 non-null  float64
 6   largura do caule     45054 non-null  float64
 7   cor do caule         45054 non-null  float64
 8   estação              45054 non-null  float64
dtypes: float64(9)
memory usage: 3.4 MB


Aqui limpamos as linhas e, por precaução, salvamos os dados em um novo csv.

In [7]:
df.to_csv('cogumelos_limpo.csv', index=False)

In [8]:
df.describe()

comestivel  diâmetro do chapéu  forma do chapéu  fixação das lamelas  \
count  45054.000000        45054.000000     45054.000000         45054.000000   
mean       0.549962          685.201447         4.725130             2.921983   
std        0.497503          399.185349         2.463306             2.633962   
min        0.000000            8.000000         0.000000             0.000000   
25%        0.000000          364.000000         2.000000             1.000000   
50%        1.000000          650.000000         6.000000             3.000000   
75%        1.000000          926.750000         6.000000             5.000000   
max        1.000000         2505.000000        10.000000            11.000000   

       cor das lamelas  altura do caule  largura do caule  cor do caule  \
count     45054.000000     45054.000000      45054.000000  45054.000000   
mean          7.326919         0.916845       1050.080370      9.453478   
std           3.202846         0.709776        782.540758      3.791821   
min           0.000000         0.000426          0.000000      0.000000   
25%           5.000000         0.341767        418.250000      6.000000   
50%           8.000000         0.767519        919.000000     11.000000   
75%          10.000000         1.328557       1523.000000     12.000000   
max          11.000000         4.461295       3569.000000     18.000000   

            estação  
count  45054.000000  
mean       1.036515  
std        0.406139  
min        0.027372  
25%        0.888450  
50%        0.943195  
75%        0.943195  
max        2.804273

In [9]:
df.nunique()

comestivel                2
diâmetro do chapéu     2032
forma do chapéu          11
fixação das lamelas      12
cor das lamelas          12
altura do caule        2358
largura do caule       3469
cor do caule             19
estação                   8
dtype: int64

Hora de dar uma olhada nos dados limpos e, ainda, contar os dados não únicos de cada coluna - aí entendemos melhor a estrutura. 

In [15]:
x = df.drop("comestivel", axis=1)
y = df["comestivel"]

Aqui separamos o resultado dos dados em X e Y.

In [18]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=42)

In [29]:
sabidinho = DecisionTreeClassifier()
sabidinho.fit(x_train, y_train)
sabidinho.score(x_test, y_test)

0.9542781045388969

Então usamos o modelo Decision Tree Classifier, inicialmente, onde conseguimos uma acurácia de cerca de 95%, ainda insuficiente.

In [30]:
from sklearn.ensemble import RandomForestClassifier
sabiducho = RandomForestClassifier()
sabiducho.fit(x_train, y_train)
sabiducho.score(x_test, y_test)

0.9806902674508934

A partir daí usamos o modelo Random Forest Classifier, que nos deu uma acurácia de 98%, suficiente para o objetivo de acurácia mínima. 